tools/tcpdrop.py

```diff
@@ -0,0 +1,222 @@
+        data4.state = state;
+        data4.tcpflags = tcpflags;
+        data4.stack_id = stack_traces.get_stackid(ctx, 0);
+        ipv4_events.perf_submit(ctx, &data4, sizeof(data4));
+
+    } else if (family == AF_INET6) {
+        struct ipv6_data_t data6 = {.pid = pid, .ip = 6};
+        bpf_probe_read(&data6.saddr, sizeof(data6.saddr),
+            sk->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
+        bpf_probe_read(&data6.daddr, sizeof(data6.daddr),
+            sk->__sk_common.skc_v6_daddr.in6_u.u6_addr32);
+        data6.dport = dport;
+        data6.sport = sport;
+        data6.state = state;
+        data6.tcpflags = tcpflags;
+        data6.stack_id = stack_traces.get_stackid(ctx, 0);
+        ipv6_events.perf_submit(ctx, &data6, sizeof(data6));
+    }
+    // else drop
+
+    return 0;
+}
+"""
+
+if debug or args.ebpf:
+    print(bpf_text)
+    if args.ebpf:
+        exit()
+
+# event data
+class Data_ipv4(ct.Structure):
+    _fields_ = [
+        ("pid", ct.c_ulonglong),
+        ("ip", ct.c_ulonglong),
+        ("saddr", ct.c_ulonglong),
+        ("daddr", ct.c_ulonglong),
+        ("sport", ct.c_ulonglong),
+        ("dport", ct.c_ulonglong),
+        ("state", ct.c_ulonglong),
+        ("tcpflags", ct.c_ulonglong),
+        ("stack_id", ct.c_ulonglong)
+    ]
+
+class Data_ipv6(ct.Structure):
+    _fields_ = [
+        ("pid", ct.c_ulonglong),
+        ("ip", ct.c_ulonglong),
+        ("saddr", (ct.c_ulonglong * 2)),
+        ("daddr", (ct.c_ulonglong * 2)),
+        ("sport", ct.c_ulonglong),
+        ("dport", ct.c_ulonglong),
+        ("state", ct.c_ulonglong),
+        ("tcpflags", ct.c_ulonglong),
+        ("stack_id", ct.c_ulonglong)
+    ]
+
+# process event
+def print_ipv4_event(cpu, data, size):
+    event = ct.cast(data, ct.POINTER(Data_ipv4)).contents
+    print("%-8s %-6d %-2d %-20s > %-20s %s (%s)" % (
+        strftime("%H:%M:%S"), event.pid, event.ip,
+        "%s:%d" % (inet_ntop(AF_INET, pack('I', event.saddr)), event.sport),
+        "%s:%s" % (inet_ntop(AF_INET, pack('I', event.daddr)), event.dport),
+        tcp.tcpstate[event.state], tcp.flags2str(event.tcpflags)))
+    for addr in stack_traces.walk(event.stack_id):
+        sym = b.ksym(addr, show_offset=True)
+        print("\t%s" % sym)
+    print("")
+
+def print_ipv6_event(cpu, data, size):
+    event = ct.cast(data, ct.POINTER(Data_ipv6)).contents
+    print("%-8s %-6d %-2d %-20s > %-20s %s (%s)" % (
+        strftime("%H:%M:%S"), event.pid, event.ip,
+        "%s:%d" % (inet_ntop(AF_INET6, event.saddr), event.sport),
+        "%s:%d" % (inet_ntop(AF_INET6, event.daddr), event.dport),
+        tcp.tcpstate[event.state], tcp.flags2str(event.tcpflags)))
+    for addr in stack_traces.walk(event.stack_id):
+        sym = b.ksym(addr, show_offset=True)
+        print("\t%s" % sym)
+    print("")
+
+# initialize BPF
+b = BPF(text=bpf_text)
+if b.get_kprobe_functions("tcp_drop"):
+    b.attach_kprobe(event="tcp_drop", fn_name="trace_tcp_drop")
```

Comment:
Checked the latest bpf-next code. I got
```
[yhs@localhost net]$ egrep -r tcp_drop
ipv4/tcp_input.c:static void tcp_drop(struct sock *sk, struct sk_buff *skb)
ipv4/tcp_input.c:			tcp_drop(sk, skb);
ipv4/tcp_input.c:		tcp_drop(sk, skb);
ipv4/tcp_input.c:		tcp_drop(sk, skb1);
ipv4/tcp_input.c:		tcp_drop(sk, skb);
ipv4/tcp_input.c:		tcp_drop(sk, rb_to_skb(node));
ipv4/tcp_input.c:	tcp_drop(sk, skb);
ipv4/tcp_input.c:	tcp_drop(sk, skb);
ipv4/tcp_input.c:			tcp_drop(sk, skb);
ipv4/tcp_input.c:		tcp_drop(sk, skb);
[yhs@localhost net]$ pwd
/home/yhs/work/bpf-next/net
[yhs@localhost net]$ uname -r
4.17.0-rc6+
[yhs@localhost net]$ 
```
Looks like tcp_drop is only for ipv4? I did not find an appropriate place to trace tcp packet drop for ipv6.

Reply:
Interesting, thanks. So I don't know yet what's going on, but IPv6 is traceable:

```
# ./tcpdrop.py
TIME     PID    IP SADDR:SPORT          > DADDR:DPORT          STATE (FLAGS)
21:07:37 7411   6  ::1:23               > ::1:53834            CLOSE (RST|ACK)
	tcp_drop+0x1
	tcp_v6_do_rcv+0x1ca
	__release_sock+0x88
	release_sock+0x30
	__inet_stream_connect+0x172
	inet_stream_connect+0x3b
	SYSC_connect+0x9e
	sys_connect+0xe
	do_syscall_64+0x73
	entry_SYSCALL_64_after_hwframe+0x3d
```

And yet if I look at tcp_v6_do_rcv(), it doesn't call tcp_drop. Maybe something is inlined?

Reply:
Just checked tcp_ipv6.c. `tcp_v6_do_rcv` calls `tcp_rcv_established` which is defined in `net/ipv4/tcp_input.c` and calls `tcp_drop`. Therefore, indeed, v6 packets could call `tcp_drop`.